agent/consul/stream/event_publisher_test.go

```diff
@@ -495,3 +540,36 @@ func TestEventPublisher_Unsubscribe_ClosesSubscription(t *testing.T) {
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "subscription was closed by unsubscribe")
 }
+
+func TestEventPublisher_Unsubscribe_FreesResourcesWhenThereAreNoSubscribers(t *testing.T) {
+	req := &SubscribeRequest{
+		Topic: testTopic,
+		Key:   "sub-key",
+	}
+
+	publisher := NewEventPublisher(newTestSnapshotHandlers(), time.Second)
+
+	sub1, err := publisher.Subscribe(req)
+	require.NoError(t, err)
+
+	// Expect a topic buffer and snapshot to have been created.
+	require.NotNil(t, publisher.topicBuffers[req.Topic][req.TopicKey()])
+	require.NotNil(t, publisher.snapCache[req.Topic][req.TopicKey()])
+
+	// Create another subscription and close the old one, to ensure the buffer and
+	// snapshot stick around as long as there's at least one subscriber.
+	sub2, err := publisher.Subscribe(req)
+	require.NoError(t, err)
+
+	sub1.Unsubscribe()
+
+	require.NotNil(t, publisher.topicBuffers[req.Topic][req.TopicKey()])
+	require.NotNil(t, publisher.snapCache[req.Topic][req.TopicKey()])
+
+	// Close the other subscription and expect the buffer and snapshot to have
+	// been cleaned up.
+	sub2.Unsubscribe()
+
+	require.Nil(t, publisher.topicBuffers[req.Topic][req.TopicKey()])
+	require.Nil(t, publisher.snapCache[req.Topic][req.TopicKey()])
+}
```

Comment:
All of these reads from the internal maps should be done holding the lock to please the race detector right?

Reply:
Yep! Good catch 😅